Add gcloud-java-compute module and immutable resources

mziccard opened the pull request:
As per title this PR adds the compute module.
This also adds classes for immutable resources (and their identities):
- DiskType
- MachineType
- License
- Region
- Zone

ajkannan commented on gcloud-java-compute/README.md:
```diff
@@ -0,0 +1,100 @@
+
+See the [Authentication](https://github.com/GoogleCloudPlatform/gcloud-java#authentication) section in the base directory's README.
+
+About Google Cloud BigQuery
```
s/BigQuery/Compute/

ajkannan commented on gcloud-java-compute/pom.xml:
```diff
@@ -0,0 +1,49 @@
+    <dependency>
+      <groupId>com.google.apis</groupId>
+      <artifactId>google-api-services-compute</artifactId>
+      <version>v1-rev93-1.21.0</version>
```
From the public docs, it looks like there's a v1 and a beta version out.  Just so that I'm on the same page, why did we choose to go with v1 over beta?

Also it looks like a new revision was released a couple days ago: [v1-rev97-1.21.0](http://search.maven.org/#artifactdetails%7Ccom.google.apis%7Cgoogle-api-services-compute%7Cv1-rev97-1.21.0%7Cjar)

mziccard replied:
We discussed with the compute team and decided to stick to v1 for now.

ajkannan commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/ComputeException.java:
```diff
@@ -0,0 +1,59 @@
+ */
+public class ComputeException extends BaseServiceException {
+
+  private static final Set<Error> RETRYABLE_ERRORS = ImmutableSet.of(new Error(500, null));
```
Is there a webpage that details which errors should be retried?  If so, it'd be good to include as a comment here.

mziccard replied:
I couldn't find it. Still trying to figure out which are the possible error codes and which of them are retryable.

ajkannan replied:
Do you mind adding a todo comment here or creating an issue for updating this set once we have that info?

mziccard replied:
Opened #662

ajkannan commented on gcloud-java-compute/README.md:
```diff
@@ -0,0 +1,100 @@
+Testing
+-------
+
+This library has tools to help make tests for code using Cloud BigQuery.
```
s/BigQuery/Compute/

ajkannan commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/DiskType.java:
```diff
@@ -0,0 +1,227 @@
+
+/**
+ * Google Compute Engine Disk type. A disk type represents the type of disk to use, such as a
+ * {@code pd-ssd} or {@code pd-standard}.
```
Could add link to https://cloud.google.com/compute/docs/reference/latest/diskTypes

ajkannan commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/License.java:
```diff
@@ -0,0 +1,98 @@
+/**
+ * A Google Compute Engine License. A License represents a software license. Licenses are used to
+ * track software usage in images, persistent disks, snapshots, and virtual machine instances.
+ */
```
Could add link to https://cloud.google.com/compute/docs/reference/latest/licenses (and same for other classes that correspond to model objects)

ajkannan commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/DiskType.java:
```diff
@@ -0,0 +1,227 @@
+
+  private static final long serialVersionUID = -944042261695072026L;
+
+  private final Long id;
```
Do we care about the deprecation status? https://cloud.google.com/compute/docs/reference/latest/diskTypes#deprecated

ajkannan commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/DiskType.java:
```diff
@@ -0,0 +1,227 @@
+
+    private Builder() {}
+
+    Builder id(Long id) {
```
Could this be private (since we have the static package scope "builder" method below)?

ajkannan commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/DiskType.java:
```diff
@@ -0,0 +1,227 @@
+
+  static DiskType fromPb(com.google.api.services.compute.model.DiskType diskTypePb) {
+    Builder builder = builder();
+    if (diskTypePb.getId() != null ) {
```
Nit: extraneous space before ')'

ajkannan commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/DiskType.java:
```diff
@@ -0,0 +1,227 @@
+      return this;
+    }
+
+    /**
```
Nit: since this builder is package scope, this javadoc is unnecessary.

ajkannan commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/DiskType.java:
```diff
@@ -0,0 +1,227 @@
+   *
+   * @see <a href="https://www.ietf.org/rfc/rfc3339.txt">RFC3339</a>
+   */
+  public String creationTimestamp() {
```
Do you think it'd be more helpful to users if we converted this into something like epoch milliseconds?

mziccard replied:
I don't have a preference. I kept the string format as that is used everywhere across the compute service but I don't mind changing.

ajkannan replied:
Do you have a preference @aozarov @mderka?

aozarov replied:
+1 for long.

ajkannan commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/DiskTypeId.java:
```diff
@@ -0,0 +1,110 @@
+
+  private final String diskType;
+
+  DiskTypeId(String project, String zone, String diskType) {
```
private?

ajkannan commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/DiskTypeId.java:
```diff
@@ -0,0 +1,110 @@
+  @Override
+  DiskTypeId setProjectId(String projectId) {
+    if (project() != null) {
+      return this;
```
I assume this means the DiskTypeId's project ID can't be changed.  Should this throw an exception instead of returning itself?

mziccard replied:
As we did in bigquery we allow the user to provide an identity for a project-owned resource without specifying the project name (e.g. `DiskTypeId.of("some-type")`). Before sending such a resource to the service we must take care of setting the project id to the default project (the one in options). The contract of `setProjectId()` is to return a new identity where the project is set (so that it can be sent to the service), so I think it should not throw.

ajkannan replied:
Ah, makes sense (I missed that this is a package scope method)

ajkannan commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/MachineType.java:
```diff
@@ -0,0 +1,298 @@
+    private Long maximumPersistentDisksSizeGb;
+
+    Builder() {}
+
```
Nit: extra line

ajkannan commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/MachineType.java:
```diff
@@ -0,0 +1,298 @@
+  private static final long serialVersionUID = -4210962597502860450L;
+
+  private final MachineTypeId machineTypeId;
+  private final Long id;
```
Do you think we should use BigInteger for IDs instead of Long (here and elsewhere)?  The protos seem to use BigInteger, and Long could lead to overflow issues.

mziccard replied:
The fields are documented to be long values in the API docs. We used this conversion also elsewhere. I would avoid BigInteger if possible

ajkannan replied:
Because it says "unsigned" in the [docs](https://cloud.google.com/compute/docs/reference/latest/machineTypes#id), I wonder if there's a chance for overflow?  It seems slightly weird that the api client library would use BigInteger if it could have used a long instead.  In the dns-alpha branch, I think there's a similar situation where I believe we chose to go with BigInteger ([docs](https://cloud.google.com/dns/api/v1/projects#number) and our [code](https://github.com/GoogleCloudPlatform/gcloud-java/blob/dns-alpha/gcloud-java-dns/src/main/java/com/google/gcloud/dns/ProjectInfo.java#L38))

mziccard replied:
You are totally right. I don't know why I was convinced it was just `long`. Thanks for spotting this!

ajkannan commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/LicenseId.java:
```diff
@@ -0,0 +1,115 @@
+    int projectsIndex = url.indexOf("/projects/");
+    int licensesIndex = url.indexOf("/global/licenses/");
+    String project = url.substring(projectsIndex + 10, licensesIndex);
+    String license = url.substring(licensesIndex + 17, url.length());
```
Is this the same thing as the [name](https://cloud.google.com/compute/docs/reference/latest/machineTypes#name) field of the model object, or is "name" something different?

mziccard replied:
It's the same

ajkannan commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/Region.java:
```diff
@@ -0,0 +1,319 @@
+    }
+  }
+
+  private final RegionId regionId;
```
Nit: here and elsewhere, can we move these variables to the beginning of the class?

ajkannan commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/License.java:
```diff
@@ -0,0 +1,98 @@
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(licenseId, chargesUseFee, selfLink);
```
To be consistent with the other classes' (eg DiskType, MachineType) hashcode methods, this could just be licenseId, right?

ajkannan commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/MachineType.java:
```diff
@@ -0,0 +1,298 @@
+  private final Integer maximumPersistentDisks;
+  private final Long maximumPersistentDisksSizeGb;
+
+  public static final class Builder {
```
This can be a package scope builder, right?

ajkannan commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/Region.java:
```diff
@@ -0,0 +1,319 @@
+    private List<ZoneId> zones;
+    private List<Quota> quotas;
+
+    Builder() {}
```
Here (and in other builders), I think we can make this constructor private.

ajkannan commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/Region.java:
```diff
@@ -0,0 +1,319 @@
+  private final List<ZoneId> zones;
+  private final List<Quota> quotas;
+
+  public static final class Builder {
```
package scope instead of public?

ajkannan commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/Region.java:
```diff
@@ -0,0 +1,319 @@
+    this.description = builder.description;
+    this.selfLink = builder.selfLink;
+    this.status = builder.status;
+    this.zones = builder.zones;
```
Should this be an immutable copy of the list?

ajkannan commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/Region.java:
```diff
@@ -0,0 +1,319 @@
+    }
+
+    Builder zones(List<ZoneId> zones) {
+      this.zones = zones;
```
Here and in quotas, should we make a copy of the list?

ajkannan commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/Zone.java:
```diff
@@ -0,0 +1,380 @@
+  private final List<MaintenanceWindow> maintenanceWindows;
+  private final RegionId region;
+
+  public static final class Builder {
```
package scope

ajkannan commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/Zone.java:
```diff
@@ -0,0 +1,380 @@
+        return false;
+      }
+      MaintenanceWindow other = (MaintenanceWindow) obj;
+      return Objects.equals(name, other.name)
```
Nit: could we use toPb() here as well?

ajkannan commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/Zone.java:
```diff
@@ -0,0 +1,380 @@
+  static Zone fromPb(com.google.api.services.compute.model.Zone zonePb) {
+    Builder builder = builder();
+    builder.zoneId(ZoneId.fromUrl(zonePb.getSelfLink()));
+    if (zonePb.getId() != null) {
```
If/when you implement a "get" with options to specify fields, will ID a field that is always returned, regardless of whether the user specifies it or not?  If it will always be returned, I think `zonePb.getId()` should never be null, and we could remove this check.  (Also applies to other `fromPb`s where we check for null ID.)

mziccard replied:
Yes I plan to have fields selector that make Id null, as we did also elsewhere

ajkannan commented on gcloud-java-compute/src/test/java/com/google/gcloud/compute/DiskTypeTest.java:
```diff
@@ -0,0 +1,73 @@
+  private static final Long DEFAULT_DISK_SIZE_GB = 10L;
+  private static final DiskTypeId DISK_TYPE_ID = DiskTypeId.of("project", "zone", "diskType");
+  private static final DiskType DISK_TYPE = DiskType.builder()
+    .id(ID)
```
Nit: should be indented 4 spaces

ajkannan commented on gcloud-java-compute/src/test/java/com/google/gcloud/compute/SerializationTest.java:
```diff
@@ -0,0 +1,157 @@
+import java.io.Serializable;
+import java.util.List;
+
+import org.junit.Test;
```
The linter complains that this should go above the "java...." imports (b/c it's a third party import).

ajkannan commented on gcloud-java-compute/src/test/java/com/google/gcloud/compute/DiskTypeIdTest.java:
```diff
@@ -0,0 +1,65 @@
+  }
+
+  @Test
+  public void testSetProjectId() {
```
Nit: could increase coverage by adding a line `assertSame(diskTypeId, diskTypeId.setProjectId("something")) `

ajkannan commented on gcloud-java-compute/src/test/java/com/google/gcloud/compute/MachineTypeTest.java:
```diff
@@ -0,0 +1,90 @@
+  }
+
+  private void compareMachineTypes(MachineType expected, MachineType value) {
+    assertEquals(expected.id(), value.id());
```
Duplicate of below.  I think you meant to remove the "id()" here to check the "equals" method.

ajkannan commented on gcloud-java-compute/src/test/java/com/google/gcloud/compute/DiskTypeTest.java:
```diff
@@ -0,0 +1,73 @@
+  }
+
+  private void compareDiskTypes(DiskType expected, DiskType value) {
+    assertEquals(expected.id(), value.id());
```
Nit: we usually add an assertEquals(expected, value) in the compare function.

ajkannan commented:
Finished first pass; looks good, mostly minor comments.

mziccard commented:
@ajkannan Thanks for the thorough pass!

I addressed all comments in 6d1f2b9 (add deprecation status and address all visibility/docs issues) and deef55c (make ID a BigInteger and put fields at the top of classes).

ajkannan commented:
It looks like Region and Zone also have deprecation status as well.  Do you think we should add a deprecation status to those two classes as well?

ajkannan commented on gcloud-java-compute/README.md:
```diff
@@ -0,0 +1,100 @@
+[![Dependency Status](https://www.versioneye.com/user/projects/56bd8ee72a29ed002d2b0969/badge.svg?style=flat)](https://www.versioneye.com/user/projects/56bd8ee72a29ed002d2b0969)
+
+-  [Homepage] (https://googlecloudplatform.github.io/gcloud-java/)
+<!-- TODO(mziccard): add link to API documentatin -->
```
Nit: this comment shows up in the README.  Also documentation is missing an "o".

ajkannan commented on gcloud-java-compute/src/test/java/com/google/gcloud/compute/DeprecationStatusTest.java:
```diff
@@ -0,0 +1,78 @@
+  private static final String DELETED = "2016-01-20T04:39:00.210-08:00";
+  private static final String DEPRECATED = "2016-01-20T04:37:00.210-08:00";
+  private static final String OBSOLETE = "2016-01-20T04:38:00.210-08:00";
+  private static final DiskTypeId DISK_TYPE_ID = DiskTypeId.of("project", "zone","diskType");
```
Lint error: missing space after comma here and in line 32.

ajkannan commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/DiskTypeId.java:
```diff
@@ -0,0 +1,124 @@
+  }
+
+  /**
+   * Returns a disk type identity given the zone disk and disk type.
```
Here and below, "zone disk" --> "zone" or "zone (as a string)" (unless "zone disk" a term used in compute that I'm unfamiliar with).

ajkannan commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/LicenseId.java:
```diff
@@ -0,0 +1,115 @@
+   *
+   * @see <a href="https://www.ietf.org/rfc/rfc1035.txt">RFC1035</a>
+   */
+  public String license() {
```
If every Resource ID subclass has a field similar to this one, do you think it'd be useful to add a field "name" and name getter to ResourceId?  I don't feel strongly either way, I'm just suggesting an alternative in case somebody has a preference.

mziccard replied:
This is similar to what we have in bigquery's identities. I don't mind changing either.

ajkannan commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/MachineType.java:
```diff
@@ -0,0 +1,326 @@
+  }
+
+  /**
+   * Returns an unique identifier for the machin type; defined by the service.
```
s/machin/machine/

ajkannan commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/RegionId.java:
```diff
@@ -0,0 +1,103 @@
+
+  @Override
+  public boolean equals(Object obj) {
+    return obj instanceof RegionId && baseEquals((RegionId) obj);
```
Here and in other ID classes' equals methods, shouldn't there also be a check that the name (here it's `region`) is equal?

mziccard replied:
`ResourceId`'s `baseEquals` compares `toUrl` that is redefined in subclasses so comparison should be already taking name into account.

ajkannan commented on gcloud-java-compute/src/test/java/com/google/gcloud/compute/DiskTypeTest.java:
```diff
@@ -0,0 +1,85 @@
+  }
+
+  @Test
+  public void testToPbAndFromPb() {
```
Up to you, but we could test that a Disk Type (or any other model object) is properly handled by to/fromPb when all optional fields are left unspecified (it would mostly catch potential NPEs).

ajkannan commented:
I'm finished with my review of this PR.

mziccard commented:
@ajkannan Comments are fixed in ea5f070 and 0940f31.

In  0465926 I add a static method to identity classes to help parse url into identity objects. It is needed also elsewhere (not in this PR) to distinguish between global and regional resources

aozarov commented on gcloud-java-compute/pom.xml:
```diff
@@ -0,0 +1,49 @@
+<?xml version="1.0"?>
```
This file is using tabs for indentation

mziccard replied:
Are you sure about this? It seems to me it's using spaces (also I just copy pasted it from an existing pom of ours).

aozarov replied:
Yes, see bellow.

aozarov commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/RegionResourceId.java:
```diff
@@ -0,0 +1,80 @@
+  @Override
+  public boolean equals(Object obj) {
+    return obj != null
+        && obj.getClass().equals(RegionResourceId.class)
```
How could this ever be true?

aozarov replied:
Same for `ZoneResourceId`

aozarov commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/DiskTypeId.java:
```diff
@@ -0,0 +1,136 @@
+    }
+  };
+
+  static final String REGEX = ZoneResourceId.REGEX + "diskTypes/[^/]+";
```
seems like this can be private (unless you were planning to use it in tests later?).
Also, you can compile it to a Pattern to make the matching slightly faster.

aozarov commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/ResourceId.java:
```diff
@@ -0,0 +1,76 @@
+  /**
+   * Base URL for a resource URL.
+   */
+  private static final String BASE_URL = "https://www.googleapis.com/compute/v1/projects/";
```
move up

aozarov commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/ResourceId.java:
```diff
@@ -0,0 +1,76 @@
+      "(https?://(www|content).googleapis.com/compute/v1/)?projects/[^/]+/";
+  private static final long serialVersionUID = -8028734746870421573L;
+
+  private String project;
```
final ?

aozarov commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/DeprecationStatus.java:
```diff
@@ -0,0 +1,164 @@
+  static <T extends ResourceId> DeprecationStatus<T> fromPb(
+      com.google.api.services.compute.model.DeprecationStatus deprecationStatusPb,
+      Function<String, T> fromUrl) {
+    return new DeprecationStatus<T>(
```
`<>`?

aozarov commented on gcloud-java-compute/src/test/java/com/google/gcloud/compute/DiskTypeIdTest.java:
```diff
@@ -0,0 +1,84 @@
+    compareDiskTypeId(diskTypeId, DiskTypeId.fromUrl(diskTypeId.toUrl()));
+    thrown.expect(IllegalArgumentException.class);
+    thrown.expectMessage("notMatchingUrl is not a valid disk type URL");
+    diskTypeId = DiskTypeId.fromUrl("notMatchingUrl");
```
Nit, redundant assignment.

aozarov commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/Zone.java:
```diff
@@ -0,0 +1,402 @@
+    }
+
+    /**
+     * Returns the disk of the maintanance window.
```
maintenance

aozarov replied:
s/disk/name/

aozarov commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/DeprecationStatus.java:
```diff
@@ -0,0 +1,164 @@
+
+  private static final long serialVersionUID = -2695077634793679794L;
+
+  private final String deleted;
```
3 values are actually timestamps. Should really present them as strings?

mziccard replied:
Timestamps are returned as RFC strings across all the service.  If we decide for one format we need to stick to it. 

I don't have a preference, I just thought to remain close to the service representation as (milli)seconds since epoch (what we use in Storage if I recall correctly) doesn't sound much more friendly than RF  string to me. I don't mind changing so if you guys have any preference I can do it. 

Also please @aozarov let me know whenever you are done reviewing so I a can address the changes alltogether

aozarov replied:
I think it is much simpler for most people to convert from millisecond epoch rather than from a string (of which you need to know the format, and much easier to get wrong). However, consider it as a nit, as I agree that neither is great and will be changed when we move to Java 8.
BTW, I would argue that consistency with the rest of *our* API is more important in that case.

Sorry for the intermediate review. Will let you when I am done (hopefully by EOD PST).

mziccard replied:
> BTW, I would argue that consistency with the rest of our API is more important in that case.

Agree! 

> Sorry for the intermediate review. Will let you know when I am done (hopefully by EOD PST).

It's fine with intermediate reviews, I just wanted to be sure to avoid pushing intermediate commits

aozarov commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/ResourceId.java:
```diff
@@ -0,0 +1,76 @@
+public abstract class ResourceId implements Serializable {
+
+  static final String REGEX =
+      "(https?://(www|content).googleapis.com/compute/v1/)?projects/[^/]+/";
```
shouldn't the base url and the regex be effected by the host in the service options?

mziccard replied:
In some APIs the resource url is returned as `https://www.googleapis.com/...` (usually `selfLink` has this format) while for other resource's fields (e.g. attached instances in disk resources) the url is returned as `https:// content.googleapis.com/...`.

aozarov replied:
Also, looks like input from the user can be partial URL. is it guarantied that input from the service will always be full URL?

mziccard replied:
Yes to parse the identity I use `selfLink` is always guaranteed to be a full URL

aozarov replied:
Can we just ignore the URL version part? A user can specify a host (e.g. when used against a beta version of the service or en emulator when/if that comes) and I am not sure if that may effect this URLs....

mziccard replied:
If you are concerned about this (and I think it's a reasonable concern) we could avoid the base url part as a whole. This will make the matching less strict but still sound enough.

aozarov replied:
Yes, I would vote for that.

aozarov commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/DeprecationStatus.java:
```diff
@@ -0,0 +1,164 @@
+
+/**
+ * The deprecation status associated to a Google Compute Engine resource.
+ */
```
`@param <T> ...`

aozarov commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/ZoneResourceId.java:
```diff
@@ -0,0 +1,80 @@
+  }
+
+  @Override
+  public boolean equals(Object obj) {
```
My preference would be:

1) not to have equals/hashCode for this abstract class but rather override baseEquals/baseHashCode to consider the zone (and call super).
2) Change ResourceId baseEquals to consider just its member variable (projectId) and not the overriden toUrl
3) Change all leaf subclass (which already check instanceof their type) to also check the equality of their specific value.

aozarov commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/DiskType.java:
```diff
@@ -0,0 +1,247 @@
+
+  private static final long serialVersionUID = -944042261695072026L;
+
+  private final BigInteger id;
```
Here in in other places with "id", do you think there is a value in returning it as a numeric value rather than a String?
I dns (e.g. `ZoneInfo`) we decided to go with String as it feels nicer to something as opaque as id.

mziccard replied:
I can use String, this will require adding a `null` check but it's ok! BTW I don't expect users to be dealing with this id very much.

aozarov replied:
+1 for String. Yes, I agree I also not expect users to use it and rather them treat it as a black box than something with meaning (e.g. numeric value).

aozarov commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/DiskType.java:
```diff
@@ -0,0 +1,247 @@
+  private final String creationTimestamp;
+  private final String description;
+  private final String validDiskSize;
+  private final String selfLink;
```
Is that going to be difference than diskTypeId.toUrl ?

mziccard replied:
Actually it's going to be the same url. We can remove it to avoid redundancy. Do you think we should keep the getter?

aozarov replied:
Did you consider renaming ResourceId.toUrl to selfLink or link? In that case I would feel more comfrotable in not having this getter.

mziccard replied:
I liked `toUrl` a little better for naming consistency with `toPb`/`fromPb`.

aozarov replied:
I like toUrl too (though I prefer toUri), but toPb/fromPb are package scope and this is public and I assume for a reason?

Because, this url is really the selfLink and because we used selfLink in other places
to describe the "Resource URI" I would vote to change it (if we want to eliminate the
redundancy) but I would not mind a  global search and replace of selfLink to toUri.

aozarov commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/DiskType.java:
```diff
@@ -0,0 +1,247 @@
+  }
+
+  /**
+   * Returns an optional textual description of the valid disk size, such as "10GB-10TB".
```
This is nasty. Why can't they have 2 fields for min and max? Is it something that we can parse?

mziccard replied:
Yes we can parse it even though there's no real documentation of which units if measure can actually be returned (mb, gb, tb). Do you think it's safe enough for us to parse it and assume that possible units are only the tree above?

aozarov replied:
I think it is worth checking that with the service owners.

mziccard replied:
I opened #683 and I'll take care of checking with the service owners.

aozarov commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/DiskType.java:
```diff
@@ -0,0 +1,247 @@
+    diskTypePb.setValidDiskSize(validDiskSize);
+    diskTypePb.setSelfLink(selfLink);
+    diskTypePb.setDefaultDiskSizeGb(defaultDiskSizeGb);
+    diskTypePb.setZone(diskTypeId.zoneId().toUrl());
```
So that redundancy (selfLink and zone fields) is also part of the underlying API...
Did you check with the service owners if that was intentional and if that may ever change?

mziccard replied:
Actually I did not. We could be building the identity out of zone+name, but selfLink has all the info we needed so I decided to use that one. This allows for more filtering. Also the `selfLink` field is always present even in the other modules.

aozarov commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/DiskTypeId.java:
```diff
@@ -0,0 +1,136 @@
+
+  /**
+   * Returns the name of the disk type resource. The name must be 1-63 characters long, and comply
+   * with RFC1035. Specifically, the name must be 1-63 characters long and match the regular
```
Do you think the constraint settings would be more useful in the setter, `of` (or in addition, or reference)?

aozarov commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/DiskTypeId.java:
```diff
@@ -0,0 +1,136 @@
+    if (!matchesUrl(url)) {
+      throw new IllegalArgumentException(url + " is not a valid disk type URL");
+    }
+    int projectsIndex = url.indexOf("/projects/");
```
Fine as is, though I feel a regexp would have been more readable (and definitely more efficient).

mziccard replied:
Yes I plan to follow your suggestion and use a Pattern

aozarov commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/LicenseId.java:
```diff
@@ -0,0 +1,127 @@
+
+  /**
+   * Returns the name of the license. The name must be 1-63 characters long, and comply with
+   * RFC1035. Specifically, the name must be 1-63 characters long and match the regular expression
```
same comment as before regarding the javadoc location of the constraint.

aozarov commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/MachineType.java:
```diff
@@ -0,0 +1,326 @@
+  }
+
+  /**
+   * Returns the number of virtual CPUs that are available to the instance.
```
I wonder why they qualify this as "guest" but don't do the same for, e.g., memory.

aozarov commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/MachineType.java:
```diff
@@ -0,0 +1,326 @@
+  /**
+   * Returns the size of all extended scratch disks assigned to the instance, defined in GB.
+   */
+  public List<Integer> scratchDisks() {
```
should we rename it to scratchDiskGb (to be consistent with maximumPersistentDisksSizeGb) ?

aozarov commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/Region.java:
```diff
@@ -0,0 +1,371 @@
+      this.usage = usage;
+    }
+
+    public String metric() {
```
javadoc here and below.

aozarov commented:
That is indeed a big API... Nice work! Done with the review. Didn't pay much attention to the tests but I assume @ajkannan did (and also looks good in coveralls).

mziccard commented:
I fixed all comments in 9d48528. I opened #683 and I suggest we keep the current String type until we close it.

aozarov commented on gcloud-java-compute/pom.xml:
```diff
@@ -0,0 +1,49 @@
+      <version>v1-rev97-1.21.0</version>
+      <scope>compile</scope>
+      <exclusions>
+	<exclusion>
```
```
\t<exclusion>
\t  <groupId>com.google.guava</groupId>
\t  <artifactId>guava-jdk5</artifactId>
\t</exclusion>
```

aozarov commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/DiskType.java:
```diff
@@ -0,0 +1,238 @@
+
+package com.google.gcloud.compute;
+
+import com.google.api.client.util.DateTime;
```
Nit, I think we should avoid depending `com.google.api.client.util` as much as possible.
The only one exception is `Data` (which is part of the API).
In this case we should use Joda (until Java8).
I see that we already use it in other [places](https://github.com/GoogleCloudPlatform/gcloud-java/search?utf8=%E2%9C%93&q=%22com.google.api.client.util%22&type=Code), so this should be done in a separate cleanup PR (even more so for Lists which should be replaced with guava).

mziccard replied:
I removed DateTime in favor of Joda's formatter in 9873ef6

aozarov commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/ResourceId.java:
```diff
@@ -0,0 +1,71 @@
+public abstract class ResourceId implements Serializable {
+
+  static final String REGEX = ".*?projects/([^/]+)/";
+  private static final String BASE_URL = "https://www.googleapis.com/compute/v1/projects/";
```
Fine as is for now but I think we can get that from the service options (without the "project/" suffix).
Maybe we should create an issue for it?

mziccard replied:
I opened #693

aozarov commented:
Looks good. I think there are 2 more issues that needs to be handled but can be handled in other PRs.
@mziccard feel free to merge if you also think that they should be done after or not at all.
